src/genetic_low_memory.c

```diff
@@ -0,0 +1,460 @@
+
+  for (unsigned int i = 0; i < dimensions; i++) {
+
+    for (unsigned int j = 0; j < UINT16_DIGIT_COUNT; j++) {
+
+      parameter[j] = child[j + dimensionIndex];
+    }
+
+    dimensionIndex += UINT16_DIGIT_COUNT;
+
+    uint32_t value = atoi(parameter);
+
+    if (value > UINT16_MAX) {
+      value = UINT16_MAX;
+    }
+
+    nextGeneration[basePositionning + i] = value;
+  }
+  (*nextGenerationSize)++;
+  basePositionning++;
+}
+
+/**
+ *
+ * We create the children based off of a uniform crossover with two parents and
+ * apply the genetic operators
+ *
+ * @param firstParent  the first selected parent solution
+ * @param secondParent  the second selected parent solution
+ * @param nextGeneration the array containing all of the created chilren
+ * @param nextGenerationSize the amount of values in the array of created
+ * children
+ */
+static void createChildren(char* firstParent, char* secondParent,
+                           uint16_t* nextGeneration,
+                           unsigned int* nextGenerationSize) {
+
+  const unsigned int arraySize = UINT16_DIGIT_COUNT * dimensions;
+  const unsigned int encodedChildStringSize = arraySize + 1;
+
+  char firstChildString[encodedChildStringSize];
+  char secondChildString[encodedChildStringSize];
+
+  firstChildString[arraySize] = '\0';
+  secondChildString[arraySize] = '\0';
+
+  for (unsigned int i = 0; i < arraySize; i++) {
+
+    if (linear_congruential_random_generator() <= 0.5) {
+
+      firstChildString[i] = firstParent[i];
+      secondChildString[i] = secondParent[i];
+    } else {
+      firstChildString[i] = secondParent[i];
+      secondChildString[i] = firstParent[i];
+    }
+  }
+
+  // We apply the mutation operator to both children
+  mutate(firstChildString, arraySize);
+  mutate(secondChildString, arraySize);
+
+  // We decode both children and add them to the next generation
+  decodeAndAddChild(nextGeneration, nextGenerationSize, firstChildString);
+  decodeAndAddChild(nextGeneration, nextGenerationSize, secondChildString);
+}
+
+/**
+ * We encode a solution as a string , the given string will always be the same
+ * length
+ *
+ * @param combinedValue the created string containing the encoded parent
+ * @param parent an array containing the parameters of the parent
+ */
+static void encode(char* combinedValue, uint16_t* parent) {
+
+  unsigned int arrIndex = 0;
+  unsigned int encodedParameterSize = UINT16_DIGIT_COUNT + 1;
+
+  for (unsigned int i = 0; i < dimensions; i++) {
+
+    char stringValue[encodedParameterSize];
+    char otherString[encodedParameterSize];
+
+    stringValue[UINT16_DIGIT_COUNT] = '\0';
+    otherString[UINT16_DIGIT_COUNT] = '\0';
+
+    unsigned int otherStringLength = 0;
+    for (int j = (UINT16_DIGIT_COUNT - 1); j >= 0; j--) {
+
+      if (parent[i] < pow(10, j)) {
+        otherString[otherStringLength] = '0';
+        otherStringLength++;
+      } else {
+        break;
+      }
+    }
```

Comment:
It depends on how hard and precise it is to compute a log in embedded C, but I don't think you need a for loop to do what you are doing since you are basically calculating a log. Something like that should work : 
```
    unsigned int numberOfZeros = (int) fmax(0, (UINT16_DIGIT_COUNT - floor(log10(parent[i])) - 1)); // Could use a define for MAX(a,b)
    char otherString[encodedParameterSize];
    memset(otherString, '0', numberOfZeros);
    otherString[numberOfZeros + 1] = '\0';

```

Reply:
Just so you know, log10 takes around 4.5K clock cycles to compute (source : https://www.nongnu.org/avr-libc/user-manual/benchmarks.html)

Reply:
I also proposed a method using only sprintf, I don't know which one is more efficient. I'll let @nicholaslegrand102 decide https://github.com/1chipML/1chipML/pull/22#discussion_r1110497642

Reply:
I decided that the sprintf implementation was more appropriate because it removed the necessity for logs